runtime/lua/vim/lsp/buf.lua

```diff
@@ -116,145 +116,92 @@ end
+    )
+    if user_choice == 0 then return nil end
+    return clients[user_choice]
   elseif #clients < 1 then
-    on_choice(nil)
+    return nil
   else
-    on_choice(clients[1])
+    return clients[1]
   end
 end
 
---- Formats the current buffer.
----
----@param options (optional, table) Can be used to specify FormattingOptions.
---- Some unspecified options will be automatically derived from the current
---- Neovim options.
---
----@see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
-function M.formatting(options)
-  local params = util.make_formatting_params(options)
-  local bufnr = vim.api.nvim_get_current_buf()
-  select_client('textDocument/formatting', function(client)
-    if client == nil then
-      return
-    end
-
-    return client.request('textDocument/formatting', params, nil, bufnr)
-  end)
-end
-
---- Performs |vim.lsp.buf.formatting()| synchronously.
----
---- Useful for running on save, to make sure buffer is formatted prior to being
---- saved. {timeout_ms} is passed on to |vim.lsp.buf_request_sync()|. Example:
----
 --- <pre>
---- autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting_sync()
+--- vim.api.nvim_command[[autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting({save = true})]]
 --- </pre>
 ---
----@param options Table with valid `FormattingOptions` entries
----@param timeout_ms (number) Request timeout
----@see |vim.lsp.buf.formatting_seq_sync|
-function M.formatting_sync(options, timeout_ms)
-  local params = util.make_formatting_params(options)
-  local bufnr = vim.api.nvim_get_current_buf()
-  select_client('textDocument/formatting', function(client)
-    if client == nil then
-      return
-    end
-
-    local result, err = client.request_sync('textDocument/formatting', params, timeout_ms, bufnr)
-    if result and result.result then
-      util.apply_text_edits(result.result, bufnr)
-    elseif err then
-      vim.notify('vim.lsp.buf.formatting_sync: ' .. err, vim.log.levels.WARN)
+---@param opts|nil
+---        * formatting_options table with valid `FormattingOptions` entries
+---        * client_id number restrict formatting to client_id
+---        * timeout_ms number Request timeout
+---        * save bool save buffer after formatting
+---@see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
+function M.formatting(opts)
+  opts = opts or {}
+  local client = vim.lsp.get_client_by_id(opts.client_id) or select_client("textDocument/formatting")
+  if client == nil then return end
+
+  local request_buf = vim.api.nvim_get_current_buf()
+  vim.api.nvim_buf_set_option(request_buf, 'modifiable', false)
+  vim.notify(string.format("Locking buffer (id: %d) while formatting.", request_buf))
+
+  local formatting_handler = function(_, result, ctx, _)
+    vim.api.nvim_buf_set_option(request_buf, 'modifiable', true)
+    if not result then return end
+    util.apply_text_edits(result, ctx.bufnr)
+    if opts.save then
```

Comment:
I'm still mixed on this. I know it is a common use case, but not sure if saving and formatting should be coupled like this. Adding a generic callback to `formatting` would be more generic. I don't realy have other use-cases in mind, so no strong opinion. Just my gut feeling that there's something off about this.

Reply:
The issue is, it becomes extremely ugly due to all the nesting:

```
```
﻿﻿vim.api.nvim_command[[
autocmd BufWritePre * lua vim.lsp.buf.formatting({callback = function()
    vim.api.nvim_buf_call(request_buf, function()
      vim.api.nvim_command("noautocmd write!") 
    end
end })]] 
```